### PR TITLE
Feature/resume

### LINKED ISSUE #72: Let ssh decide its control sockets etc
Hi. I have a project where I'm trying to control ssh(&maybe sftp) from a Rust parent process, and I'd love to use your library.

However, the library seems very opinionated about passing `-S` etc options.

How would you feel about an API that lets `ssh` just do what it is configured to do, instead of forcing its hand so much? Right now, the library won't e.g. let me reuse a master connection that was opened from elsewhere.

### PR DESCRIPTION
**Pull request by NobodyXu**
Fixed #72 

This PR adds:
 - `Session::leak`
 - `Session::resume`
 - `Session::resume_mux`
 - `Session::force_terminate`

and integration tests for them.

This PR also added a minor change to `.github/workflows/style.yml` to make sure it failed on warnings like missing items and updated `run_ci_tests.sh` to remove temp dirs.

Signed-off-by: Jiahao XU <Jiahao_XU@outlook.com>

Edit:

It also [Fix random error in step "Test ssh connectivity"](https://github.com/openssh-rust/openssh/pull/80/commits/39298e7ce884be8175af6a74d58cef9bc81eae3a).

### CODE REVIEW

**Comment by jonhoo** [comment]
This change is [Reviewable](https://reviewable.io/reviews/openssh-rust/openssh/80)

**Comment by NobodyXu** [comment]
> # [Codecov](https://codecov.io/gh/openssh-rust/openssh/pull/80?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=None) Report
> 
> > Merging [#80](https://codecov.io/gh/openssh-rust/openssh/pull/80?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=None) ([bd85a0b](https://github.com/openssh-rust/openssh/commit/bd85a0bc6c62e0900a04e36aee1f4eec06cf98c0)) into [master](https://codecov.io/gh/openssh-rust/openssh/commit/13e80b7496bea9f40d27806a6b7ac84edfc3b867?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=None) ([13e80b7](https://github.com/openssh-rust/openssh/commit/13e80b7496bea9f40d27806a6b7ac84edfc3b867)) will **increase** coverage by `0.58%`.
> > The diff coverage is `97.29%`.
> 
> [Impacted Files](https://codecov.io/gh/openssh-rust/openssh/pull/80?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=None) 	Coverage Δ 	
> [src/native_mux_impl/session.rs](https://codecov.io/gh/openssh-rust/openssh/pull/80/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=None#diff-c3JjL25hdGl2ZV9tdXhfaW1wbC9zZXNzaW9uLnJz) 	`93.75% <96.15%> (+0.52%)` 	⬆️
> [src/process_impl/session.rs](https://codecov.io/gh/openssh-rust/openssh/pull/80/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=None#diff-c3JjL3Byb2Nlc3NfaW1wbC9zZXNzaW9uLnJz) 	`87.02% <97.22%> (+1.58%)` 	⬆️
> [src/session.rs](https://codecov.io/gh/openssh-rust/openssh/pull/80/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=None#diff-c3JjL3Nlc3Npb24ucnM=) 	`74.28% <100.00%> (+5.32%)` 	⬆️

So happy to finally see the cov report in PR!

File: .github/workflows/style.yml [hunk 1]
```diff
@@ -55,4 +55,4 @@ jobs:
           command: doc
           args: --no-deps --all-features
         env:
-          RUSTDOCFLAGS: --cfg docsrs
+          RUSTDOCFLAGS: --cfg docsrs -Dwarnings
```
Comment (jonhoo) [inline .github/workflows/style.yml:58]: I'm not a fan of denying warnings, even in CI because new warnings will _always_ accumulate over time simply as they get added to rustc/rustdoc/clippy. This is why we have GitHub annotations for PRs, which indicate the warnings without blocking the PR.

File: src/session.rs [hunk 2]
```diff
@@ -65,6 +65,31 @@ impl From<native_mux_impl::Session> for Session {
...
+    ///
+    /// This connects to the ssh multiplex master using process mux impl.
+    #[cfg(feature = "process-mux")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "process-mux")))]
+    pub fn resume(ctl: Box<Path>, master_log: Option<Box<Path>>) -> Self {
+        process_impl::Session::resume(ctl, master_log).into()
+    }
+
+    /// Same as [`Session::resume`] except that it connects to
+    /// the ssh multiplex master using process mux impl.
```
Comment (jonhoo) [inline src/session.rs:86]: I think this should say "native mux"?

File: src/session.rs [hunk 3]
```diff
@@ -295,4 +320,16 @@ impl Session {
     pub async fn close(self) -> Result<(), Error> {
         delegate!(self.0, imp, { imp.close().await })
     }
+
+    /// Leak the underlying ssh multiplex master.
+    /// Return (path to control socket, path to ssh multiplex output log)
+    pub fn leak(self) -> (Box<Path>, Option<Box<Path>>) {
```
Comment (jonhoo) [inline src/session.rs:326]: I think we should avoid calling this "leak", since that has a pretty strong association with leaking memory in Rust (see `Box::leak`). One option is to instead just have a configuration method for "what should happen on drop" — something like `close_on_drop(yes: bool)`, which the user can toggle on or off. If you want to keep it a destructing method like this, I'd go with `fn detach`.

File: src/session.rs [hunk 4]
```diff
@@ -295,4 +320,16 @@ impl Session {
...
+        delegate!(self.0, imp, { imp.leak() })
+    }
+
+    /// Force terminate the remote connection, even if it is created
+    /// by [`Session::resume`] or [`Session::resume_mux`].
+    pub async fn force_terminate(self) -> Result<(), Error> {
```
Comment (jonhoo) [inline src/session.rs:332]: Here, too, I think we can make the API a little clearer. One option is to rely on the same kind of configuration as mentioned above (`close_on_drop(yes: bool)`), where that defaults to no for resumed sessions, but can be set to yes. Alternatively, I think it's reasonable to make an explicit call to `close` _always_ mean `-O exit` (including for resumed sessions), and to _just_ change the default behavior of drop depending on whether the session was resumed or not.

**Review by jonhoo — CHANGES_REQUESTED** [review]

File: .github/workflows/minimal.yml [hunk 5]
```diff
@@ -34,6 +34,7 @@ jobs:
           ssh -i .test-key -v -p 2222 -l test-user 127.0.0.1 -o StrictHostKeyChecking=accept-new -o UserKnownHostsFile=/tmp/openssh-rs/known_hosts whoami
         name: Test ssh connectivity
       - run: |
+          sleep 30
```
[the remark this answers is not in the document]
Comment (NobodyXu) [reply .github/workflows/minimal.yml:37]: I added this because the ci randomly failed.

The `openssh-server` container actually calls `apk` to install some softwares required to fetch the mod `DOCKER_MODS: linuxserver/mods:openssh-server-ssh-tunnel` and then downloads and applies the mod on the fly before starting `sshd` process, so the startup time varies and sometimes it just falls.

I tried `sleep 5` but it still failed so I changed that to `sleep 30`.

I guess I can try `sleep 15` as 15s should be enough.

[the remark this answers is not in the document]
Comment (NobodyXu) [reply .github/workflows/minimal.yml]: Done!

File: .github/workflows/style.yml (hunk 1 above)
Comment (NobodyXu) [reply .github/workflows/style.yml:58]: Done!

File: src/session.rs (hunk 2 above)
Comment (NobodyXu) [reply src/session.rs]: Done!

File: src/session.rs (hunk 3 above)
Comment (NobodyXu) [reply src/session.rs:326]: I'd like to keep it as it return the path to control socket and the path to the master_log, which is useful for process mux implementation.

Comment (NobodyXu) [reply src/session.rs]: Done!

File: src/session.rs (hunk 4 above)
Comment (NobodyXu) [reply src/session.rs]: Done!

File: tests/openssh.rs [hunk 6]
```diff
@@ -829,6 +829,96 @@ async fn local_socket_forward() {
...
+        let (ctl2, master_log2) = session2.detach();
+
+        // Second resume_mux to ensure close handles tempdir set to None correctly
+        let session3 = Session::resume_mux(ctl2, master_log2);
+        session3.check().await.unwrap();
+
+        session3.close().await.unwrap();
+    }
+
+    // test force_terminate
```
Comment (jonhoo) [inline tests/openssh.rs:902]: This doesn't seem accurate any more since there's no `force_terminate` now.

File: src/session.rs [hunk 7]
```diff
@@ -65,6 +65,31 @@ impl From<native_mux_impl::Session> for Session {
...
 impl Session {
+    /// Resume the connection using path to control socket and
+    /// path to ssh multiplex output log.
+    ///
+    /// If you do not use `-E` option (or redirection) to write
+    /// the log of the ssh multiplex master to the disk, you can
+    /// simply pass `None` to `master_log`.
+    ///
+    /// [`Session`] created this way will not be terminated on drop,
+    /// but can be forced terminated by [`Session::force_terminate`].
```
Comment (jonhoo) [inline src/session.rs:76]: This should be called `close` now I think.

Comment (jonhoo) [reply src/session.rs:76]: We should also update the documentation on `close` to mention that it will terminate the control master even if the session was resumed.

**Review by jonhoo — CHANGES_REQUESTED** [review]
Nice, almost there! I like the simplicity of where we ended up. Just some cleanup related to old references to `force_terminate` left.

File: src/session.rs (hunk 7 above)
Comment (NobodyXu) [reply src/session.rs]: Done!

File: tests/openssh.rs (hunk 6 above)
Comment (NobodyXu) [reply tests/openssh.rs]: Done!

File: .github/workflows/minimal.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (Huanzo) [reply .github/workflows/minimal.yml]: To maybe get rid of a hard coded sleep one could use `ssh-keyscan` to probe the ssh server. I use `until ssh-keyscan "$IP" > /root/.ssh/known_hosts; do sleep 1; done` in one of my own CIs to wait for the remote host to accept ssh connections.

[the remark this answers is not in the document]
Comment (NobodyXu) [reply .github/workflows/minimal.yml]: Nice!
I will have a try.

[the remark this answers is not in the document]
Comment (NobodyXu) [reply .github/workflows/minimal.yml]: @Huanzo Thank you very much!
I've now fixed it by using a proper solution, combining your solution with a `timeout`.

File: src/session.rs [hunk 8]
```diff
@@ -292,7 +317,17 @@ impl Session {
     }
 
     /// Terminate the remote connection.
+    /// It would terminate the ssh multiplex server
+    /// regardless of how it is created.
```
Comment (jonhoo) [inline src/session.rs:321]:
```suggestion
    ///
    /// This destructor terminates the ssh multiplex server
    /// regardless of how it was created.
```

File: .github/workflows/test.yml [hunk 9]
```diff
@@ -30,6 +30,7 @@ jobs:
       - uses: actions/checkout@v2
       - run: |
           # Wait for startup of openssh-server
+          timeout 15 ./wait_for_sshd_start_up.sh
           sleep 15
```
Comment (jonhoo) [inline .github/workflows/test.yml:34]: Did you mean to remove the `sleep 15` here too?

**Review by jonhoo — CHANGES_REQUESTED** [review]
Another two changes and I think we're there!

File: .github/workflows/test.yml (hunk 9 above)
Comment (NobodyXu) [reply .github/workflows/test.yml:34]: Oops, good catch!

Comment (NobodyXu) [reply .github/workflows/test.yml]: Done!

**Review by jonhoo — APPROVED** [review]
Great, thank you!

**Comment by jonhoo** [comment]
Released in 0.9.3 :tada:
